lib/src/component_declaration/function_component.dart

```diff
@@ -0,0 +1,181 @@
+///     return Dom.div()(props.items);
+///   },
+///   null,
+///   propsFactory: PropsFactory.fromUiFactory(Foo),
+///   displayName: 'Bar',
+/// );
+/// ```
+///
+/// __OR__ Set [config] to `null` when using `UiProps`.
+///
+/// ```dart
+/// UiFactory<UiProps> Baz = uiFunction((props) {
+///     return Dom.div()('prop id: ${props.id}');
+///   },
+///   null,
+///   displayName: 'Baz',
+/// );
+/// ```
+///
+/// Learn more: <https://reactjs.org/docs/components-and-props.html#function-and-class-components>.
+// FIXME right now only top level factory declarations will generate props configs.
+UiFactory<TProps> uiFunction<TProps extends UiProps>(
+  UiFunctionComponent<TProps> functionComponent,
+  FunctionComponentConfig<TProps> config, {
+  PropsFactory<TProps> propsFactory,
+  String displayName,
+}) {
+  if (config != null) {
+    if (propsFactory != null) {
+      throw ArgumentError('propsFactory cannot be used along with config');
+    }
+    propsFactory = config.propsFactory;
+    displayName ??= config.componentName;
+  }
+
+  // Get the display name from the inner function if possible so it doesn't become `_uiFunctionWrapper`
+  displayName ??= _getFunctionName(functionComponent);
+
+  dynamic _uiFunctionWrapper(Map props) {
+    return functionComponent(propsFactory.jsMap(props as JsBackedMap));
+  }
+
+  ReactDartFunctionComponentFactoryProxy factory =
+      react.registerFunctionComponent(
+    _uiFunctionWrapper,
+    displayName: displayName,
+  );
+
+  if (propsFactory == null) {
+    if (TProps != UiProps && TProps != _GenericUiProps) {
+      throw ArgumentError(
+          'config.propsFactory must be provided when using custom props classes');
+    }
+    propsFactory = PropsFactory.fromUiFactory(
+            ([backingMap]) => _GenericUiProps(factory, backingMap))
+        as PropsFactory<TProps>;
+  }
+
+  TProps _uiFactory([Map backingMap]) {
+    TProps builder;
+    if (backingMap == null) {
+      builder = propsFactory.jsMap(JsBackedMap());
+    } else if (backingMap is JsBackedMap) {
+      builder = propsFactory.jsMap(backingMap);
+    } else {
+      builder = propsFactory.map(backingMap);
+    }
+
+    return builder..componentFactory = factory;
+  }
+
+  return _uiFactory;
+}
+
+String _getFunctionName(Function function) {
+  return getProperty(function, 'name') ?? getProperty(function, '\$static_name');
+}
+
+class _GenericUiProps extends UiProps {
+  @override
+  final Map props;
+
+  _GenericUiProps(ReactComponentFactoryProxy componentFactory, [Map props])
+      : this.props = props ?? JsBackedMap() {
+    this.componentFactory = componentFactory;
+  }
+
+  @override
+  String get propKeyNamespace => '';
+
+  @override
+  bool get $isClassGenerated => true;
+}
+
+/// Helper class used to keep track of generated information for [uiFunction].
+@protected
+class FunctionComponentConfig<TProps extends UiProps> {
+  @protected
+  final PropsFactory<TProps> propsFactory;
+  final String componentName;
```

Comment:
How is this different from `displayName`? Or can the builder not always generate it?

Reply:
It's not different; we should probably just rename it to `displayName`

Reply:
Renamed this in cec79b9